pkg/controller/zerocapacity/zero_controller.go

```diff
@@ -0,0 +1,477 @@
+}
+
+func (z *Controller) execute(httpClient http.HttpClient, request reconcile.Request, instance Resource) (reconcile.Result, error) {
+	if !z.isZeroPodReady(request, instance) {
+		// Don't requeue as reconcile request is received when the zero pod becomes ready
+		return reconcile.Result{}, nil
+	}
+
+	if err := instance.Exec(httpClient); err != nil {
+		z.Log.Error(err, "Unable to execute action on zero-capacity pod", "request.Name", request.Name)
+		return reconcile.Result{}, instance.UpdatePhase(ZeroFailed, err)
+	}
+
+	return reconcile.Result{}, instance.UpdatePhase(ZeroRunning, nil)
+}
+
+func (z *Controller) waitForExecutionToComplete(httpClient http.HttpClient, request reconcile.Request, instance Resource) (reconcile.Result, error) {
+	phase, err := instance.ExecStatus(httpClient)
+
+	if err != nil || phase == ZeroFailed {
+		z.Log.Error(err, "Execution failed", "request.Name", request.Name)
+		return reconcile.Result{}, instance.UpdatePhase(ZeroFailed, err)
+	}
+
+	if phase == ZeroSucceeded {
+		return reconcile.Result{}, instance.UpdatePhase(ZeroSucceeded, nil)
+	}
+
+	// Execution has not completed, or it's state is unknown, wait 1 second before retrying
+	return reconcile.Result{
+		Requeue:      true,
+		RequeueAfter: 1 * time.Second,
+	}, nil
+}
+
+func (z *Controller) cleanupResources(request reconcile.Request) (reconcile.Result, error) {
+	ctx := context.Background()
+	meta := metav1.ObjectMeta{
+		Namespace: request.Namespace,
+		Name:      request.Name,
+	}
+	var allErrors error
+
+	// Delete the zero-capacity pod so that it leaves the Infinispan cluster
+	if err := z.Delete(ctx, &corev1.Pod{ObjectMeta: meta}); err != nil && !errors.IsNotFound(err) {
+		allErrors = fmt.Errorf("Unable to delete zero-capacity pod: %w", err)
+	}
+
+	// Delete the configmap as it's no longer required
+	if err := z.Delete(ctx, &corev1.ConfigMap{ObjectMeta: meta}); err != nil && !errors.IsNotFound(err) {
+		allErrors = wrapErr(allErrors, fmt.Errorf("Unable to delete configMap: %w", err))
+	}
+	return reconcile.Result{}, nil
+}
+
+func wrapErr(old, new error) error {
+	if old != nil {
+		return fmt.Errorf("%w; %s", old, new.Error())
+	}
+	return new
+}
+
+func (z *Controller) isZeroPodReady(request reconcile.Request, instance Resource) bool {
+	pod := &corev1.Pod{}
+	key := types.NamespacedName{
+		Name:      request.Name,
+		Namespace: request.Namespace,
+	}
+
+	if err := z.Get(context.Background(), key, pod); err != nil {
+		return false
+	}
+
+	for _, c := range pod.Status.Conditions {
+		if c.Type == corev1.PodReady && c.Status == corev1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
+func (z *Controller) zeroPodSpec(name, namespace string, configMap *corev1.ConfigMap, ispn *v1.Infinispan, zeroSpec *Spec,
+	podSpec corev1.PodSpec) *corev1.Pod {
+
+	// Utilise the provided PodSpec to retrieve the identities vol
+	var identitiesVol corev1.Volume
+	for _, v := range podSpec.Volumes {
+		if v.Name == ispnCtrl.IdentitiesVolumeName {
```

Comment:
This is static config always. We can add this and get Secret name from the Infinispan CR.